sdk/identity/src/token_credentials/default_credentials.rs

```diff
@@ -51,38 +88,99 @@ impl DefaultAzureCredentialBuilder {
         self
     }
 
-    /// Create a `DefaultAzureCredential` from this builder.
-    pub fn build(&self) -> DefaultAzureCredential {
-        let source_count = usize::from(self.include_azure_cli_credential)
-            + usize::from(self.include_azure_cli_credential)
-            + usize::from(self.include_managed_identity_credential);
-        let mut sources = Vec::<DefaultAzureCredentialEnum>::with_capacity(source_count);
+    /// Get a list of the credential types to include.
+    fn included(&self) -> Vec<DefaultAzureCredentialType> {
+        let mut sources = Vec::new();
+        if self.include_specific_credential {
+            sources.push(DefaultAzureCredentialType::Specific);
+        }
         if self.include_environment_credential {
-            sources.push(DefaultAzureCredentialEnum::Environment(
-                super::EnvironmentCredential::default(),
-            ));
+            sources.push(DefaultAzureCredentialType::Environment);
+        }
+        if self.include_app_service_managed_identity_credential {
+            sources.push(DefaultAzureCredentialType::AppService);
         }
-        if self.include_managed_identity_credential {
-            sources.push(DefaultAzureCredentialEnum::ManagedIdentity(
-                ImdsManagedIdentityCredential::default(),
-            ));
+        if self.include_virtual_machine_managed_identity_credential {
+            sources.push(DefaultAzureCredentialType::VirtualMachine);
         }
         if self.include_azure_cli_credential {
-            sources.push(DefaultAzureCredentialEnum::AzureCli(
-                AzureCliCredential::new(),
-            ));
+            sources.push(DefaultAzureCredentialType::AzureCli);
+        }
+        sources
+    }
+
+    fn try_create(
+        &self,
+        included: &Vec<DefaultAzureCredentialType>,
+    ) -> Vec<DefaultAzureCredentialEnum> {
+        // If specific credential is included, try to create it.
+        // Use only the specific credential if it is created successfully.
+        if self.include_specific_credential {
+            if let Ok(credential) = SpecificAzureCredential::create(self.options.clone()) {
+                return vec![DefaultAzureCredentialEnum::Specific(credential)];
+            }
         }
+
+        let mut sources = Vec::<DefaultAzureCredentialEnum>::with_capacity(included.len());
+        for source in included {
+            match source {
+                DefaultAzureCredentialType::Specific => {}
```

Comment:
It is not. It is an idea that I introduced in this SDK.

Reply:
It's my understanding unless something makes us more rust-like, we should model what exists in other SDKs, of which this would implement a rather different experience.

CC: @heaths

Reply:
The experience is different only when you opt into setting `AZURE_CREDENTIAL_TYPE`. For example, deploying to App Service, I can set it to `appservice` and the corresponding managed identity will be used. For a VM deployment, `virtualmachine` may be set and the corresponding managed identity will be used. Locally for development, `azurecli` may be set and it is guaranteed to use it. There is not delay in checking other credential types. Performance and low level flexibility is more rust-like.

Reply:
Part of the ideals for the SDKs is to align implementations with other SDKs except where it's making it more idiomatic for the language.  "Try" verses "Result" is an idiomatic difference.

This isn't a language-idiomatic difference, but a unique feature.  (hence my CCing heaths)

Reply:
I started an [internal design discussion](https://teams.microsoft.com/l/message/19:f6d52ac6465c40ea80dc86b8be3825aa@thread.skype/1704231889101?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=3e17dcb0-4257-4a30-b843-77f47f1d4121&parentMessageId=1704231889101&teamName=Azure%20SDK&channelName=General&createdTime=1704231889101).

Reply:
The sentiment during review was to not add `SpecificAzureCredential` to `DefaultAzureCredential`, so I will remove it here.